app/femr/data/daos/system/PatientRepository.java

```diff
@@ -351,17 +353,37 @@ public IPatient retrievePatientById(Integer id) {
     @Override
     public IPatient savePatient(IPatient patient) {
 
-        IPatient response = null;
-        try {
-
-            Ebean.save(patient);
-        } catch (Exception ex) {
+        Optional<Patient> maxIdPatient;
+        Boolean savedSafely = false;
+        while (!(savedSafely)) {
+            try {
+                maxIdPatient = QueryProvider.getPatientQuery()
+                        .where().eq("kit_id", patient.getPatientKey().getKitId())
+                        .orderBy("id DESC")
+                        .setMaxRows(1)
+                        .findOneOrEmpty();
+
+                if (maxIdPatient.isPresent()) {
+                    patient.setId(maxIdPatient.get().getId() + 1);
+                } else {
+                    patient.setId(1);
+                }
+                Ebean.save(patient);
+                savedSafely = true;
+            } catch (io.ebean.DuplicateKeyException ex) {
+                // id must have just been taken, try again
+                System.out.println("duplicate key exception, try to get a valid id again");
+            }
+            catch (Exception ex) {
 
-            //is it necessary to pass all details about object in log?
-            Logger.error("PatientRepository-savePatient", ex.getMessage());
-            throw ex;
+                //is it necessary to pass all details about object in log?
+                Logger.error("PatientRepository-savePatient", ex.getMessage());
+                throw ex;
+            }
         }
 
+        //io.ebean.DuplicateKeyException
+
         return patient;
```

Comment:
general note for this file, there is a get patient by id function. This may now return more than one patient since they are not uniquely identified by id anymore. There should also be a function for get patient by kek